Update Crunchy PostgreSQL Operator to v4.2.0

cbandy opened the pull request:
Signed-off-by: Chris Bandy <chris.bandy@crunchydata.com>

Thanks submitting your Operator. Please check below list before you create your Pull Request.
*************************************************
**Flat operator directory structure is obsolete from 23-rd of October 2019, only nested directory structure will be accepted.**
*************************************************

### New Submissions

* [x] Has you operator [nested directory structure](https://github.com/operator-framework/community-operators/blob/master/docs/contributing.md#create-a-bundle)?
* [ ] Have you selected the Project *Community Operator Submissions* in your PR on the right-hand menu bar?
* [x] Are you familiar with our [contribution guidelines](https://github.com/operator-framework/community-operators/blob/master/docs/contributing.md)?
* [x] Have you [packaged and deployed](https://github.com/operator-framework/community-operators/blob/master/docs/testing-operators.md) your Operator for Operator Framework?
* [x] Have you tested your Operator with all Custom Resource Definitions?
* [x] Have you tested your Operator in all supported [installation modes](https://github.com/operator-framework/operator-lifecycle-manager/blob/master/doc/design/building-your-csv.md#operator-metadata)?
* [x] Is your submission [signed](https://github.com/operator-framework/community-operators/blob/master/docs/contributing.md#sign-your-work)?

### Updates to existing Operators

* [ ] Is your new CSV pointing to the previous version with the `replaces` property?
* [x] Is your new CSV referenced in the [appropriate channel](https://github.com/operator-framework/community-operators/blob/master/docs/contributing.md#bundle-format) defined in the `package.yaml` ?
* [x] Have you tested an update to your Operator when deployed via OLM?
* [x] Is your submission [signed](https://github.com/operator-framework/community-operators/blob/master/docs/contributing.md#sign-your-work)?

### Your submission should not

* [x] Modify more than one operator
* [x] Modify an Operator you don't own
* [x] Rename an operator - please remove and add with a different name instead
* [x] Submit operators to both `upstream-community-operators` and `community-operators` at once
* [x] Modify any files outside the above mentioned folders
* [x] Contain more than one commit. **Please squash your commits.**

### Operator Description must contain (in order)

1. [x] Description about the managed Application and where to find more information
2. [x] Features and capabilities of your Operator and how to use it
3. [x] Any manual steps about potential pre-requisites for using your Operator

### Operator Metadata should contain

* [x] Human readable name and 1-liner description about your Operator
* [x] Valid [category name](https://github.com/operator-framework/community-operators/blob/master/docs/required-fields.md#categories)<sup>1</sup>
* [x] One of the pre-defined [capability levels](https://github.com/operator-framework/operator-courier/blob/4d1a25d2c8d52f7de6297ec18d8afd6521236aa2/operatorcourier/validate.py#L556)<sup>2</sup>
* [x] Links to the maintainer, source code and documentation
* [x] Example templates for all Custom Resource Definitions intended to be used
* [x] A quadratic logo

Remember that you can preview your CSV [here](https://operatorhub.io/preview).

--

<sup>1</sup> If you feel your Operator does not fit any of the pre-defined categories, file a PR against this repo and explain your need

<sup>2</sup> For more information see [here](https://github.com/operator-framework/operator-sdk/blob/master/doc/images/operator-capability-level.svg)

robszumski commented:
The Crunchydata has exhibited a ton of new auto-remediation and auto-healing that represent an "auto pilot" level in the capability model. I propose we include that as part of this PR.

[Detailed release notes here](https://github.com/CrunchyData/postgres-operator/releases/tag/v4.2.0), but this is the relevant details:

> Failed primaries now automatically heal, which significantly reduces the time in which they can rejoin the cluster.

> A major change to this container is that the PostgreSQL process is now managed by Patroni. This allows a PostgreSQL cluster that is deployed by the PostgreSQL Operator to manage its own uptime and availability, to elect a new leader in the event of a downtime scenario, and to automatically heal after a failover event.

In addition, Pod anti-affinity and backups with retention policies make this a really exemplary Operator.

cbandy commented:
Thanks, @robszumski! I've pushed the change in amended commit a61922a.

komish commented:
@cbandy Just some small notes I had from testing the previous version, figured now is as good a time as any to get these updated. These probably also apply to the community-operators version of this PR https://github.com/operator-framework/community-operators/pull/1107 as well as the CSV generation tooling you might be using.

---

- [ ] I believe the tagging in the postgres-operator repository has changed format to `vX.Y.Z`. Does this documentation line need to be updated?

https://github.com/operator-framework/community-operators/blob/e43847669bbf3b124bd445cbb2e44e675a0eca9b/community-operators/postgresql/4.2.0/postgresoperator.v4.2.0.clusterserviceversion.yaml#L138

- [ ] The command here has an extra backslash.

https://github.com/operator-framework/community-operators/blob/e43847669bbf3b124bd445cbb2e44e675a0eca9b/community-operators/postgresql/4.2.0/postgresoperator.v4.2.0.clusterserviceversion.yaml#L217

- [ ] `PGO_AUTH_SERVER` here renders the word AUTH with italics on dashboard. I believe the underscores need to be escaped. or the variable needs to be in backticks.

https://github.com/operator-framework/community-operators/blob/e43847669bbf3b124bd445cbb2e44e675a0eca9b/community-operators/postgresql/4.2.0/postgresoperator.v4.2.0.clusterserviceversion.yaml#L264

- [ ] Does the CSV need to include a "replaces" field here?

cbandy commented:
Thank you for the quick review!

> * [ ]  I believe the tagging in the postgres-operator repository has changed format to `vX.Y.Z`. Does this documentation line need to be updated?

Yes, it does. Fixed.

> * [ ]  The command here has an extra backslash.

Fixed.

> * [ ]  `PGO_AUTH_SERVER` here renders the word AUTH with italics on dashboard. I believe the underscores need to be escaped. or the variable needs to be in backticks.

Fixed.

> * [ ]  Does the CSV need to include a "replaces" field here?

As I understand it, the `replaces` field indicates that OLM can safely perform an automated upgrade. We've omitted the field because this version does require some intervention by an administrator. There are sections in the documentation outlining what's required depending on one's current settings and version:

https://access.crunchydata.com/documentation/postgres-operator/latest/upgrade/

J0zi commented:
@cbandy, please get [signature](https://github.com/operator-framework/community-operators/blob/master/docs/contributing.md#sign-your-work) from @jkatz as he is the author.

komish commented:
> Thank you for the quick review!
> 
> > * [ ]   I believe the tagging in the postgres-operator repository has changed format to `vX.Y.Z`. Does this documentation line need to be updated?
> 
> Yes, it does. Fixed.
> 
> > * [ ]   The command here has an extra backslash.
> 
> Fixed.
> 
> > * [ ]   `PGO_AUTH_SERVER` here renders the word AUTH with italics on dashboard. I believe the underscores need to be escaped. or the variable needs to be in backticks.
> 
> Fixed.
> 
> > * [ ]   Does the CSV need to include a "replaces" field here?
> 
> As I understand it, the `replaces` field indicates that OLM can safely perform an automated upgrade. We've omitted the field because this version does require some intervention by an administrator. There are sections in the documentation outlining what's required depending on one's current settings and version:
> 
> https://access.crunchydata.com/documentation/postgres-operator/latest/upgrade/

Understood. Thanks for the quick work there!

cbandy commented:
> @cbandy, please get [signature](https://github.com/operator-framework/community-operators/blob/master/docs/contributing.md#sign-your-work) from @jkatz as he is the author.

Fixed in 5bc4110.
